### ISSUE TITLE
Support IP Address CIDR/Subnets

### ISSUE THREAD

**Issue by sparrell** [oasis-tcs/openc2-oc2ls#134]
LS-008 (https://lists.oasis-open.org/archives/openc2-comment/201811/msg00007.html):
> Earlier CSDs supported (via example at least) the ability to provide CIDR notation for IP addresses. The proposed IP serialization in the CD removes this option. Supporting the CIDR capability is important, especially in proactive scenarios.

**Comment by sparrell** [comment]
In response to https://lists.oasis-open.org/archives/openc2-comment/201811/msg00007.html email https://lists.oasis-open.org/archives/openc2-comment/201811/msg00009.html proposed:

`Suggest the following substantive change to the "Open Command and Control (OpenC2) Profile for Stateless Packet Filtering Version 1.0"  

BACKGROUND: 
As currently written the 'ip_addr' target in section 2.1.2.1 is not consistent with the Language Specification and if corrected to be consistent with the language specification, then the SLPF will no longer support CIDR notation.

PROPOSED CHANGE: 
Add a new target to the language specification to accommodate CIDR notation for IPv4 and IPv6 addresses  
Modify table 2.1.2-1 in the SLPF to include the new targets.
Update table 2.3-1 to in the SLPF to include the new targets in the command matrix 
Add new subsection to section 2.3.1 to indicate that the new target is appropriate for the 'allow' action
Update table 3-1 in the SLPF to include the commands affected by the new target
Update the conformance targets in section 3 of the SLPF to reflect the new target

JUSTIFICATION: 
The Technical Committee has identified use cases such that the ability to allow (or deny) a range of addresses with a single command is warranted.  `

**Comment by alevere** [comment]
I tentatively agree, once we have some guidance from the language spec, I would like to see the pull request for this one. Probably need to include the deny command too. I would like it to be allowed to support only a /32- some systems operate this way they expect you to explode the CIDR first.

**Comment by jmbrule** [comment]
This is an FYI:  A pull request was made in the SLPF to accommodate an 'ip_net' target so that deny and allow could be applied to a range of addresses.  The pull request will be modified in accordance with the consensus reached.  As of now, the SLPF and the lang spec are not consistent

**Comment by romanojd** [comment]
The consensus from the January F2F was to redefine the IP Address from a binary to a string to allow for CIDR notation. The other proposals to add additional targets or add additional properties to the existing targets were all found to be overly, and unnecessarily complex. In the end the simplier, flatter solution gained the consensus.

**Comment by davaya** [comment]
I disagree strongly that the consensus was to change the type of IP Address from Binary to String.   The consensus was that for JSON serialization:
* IP Address shall be serialized as a String with the expected value ("1.2.3.4")
* An IP Net target is still required, because in some cases (such as the value that goes in an IP packet's source and destination address, a CIDR expression is not valid)
* IP Net target shall be serialized as a String with the expected value ("1.2.3.4" or "1.2.3.4/16"), i.e., the prefix portion of the value is optional.
* Separate IPv4-Addr and IPv6-Addr types replace the merged IP-Addr type because the values are serialized differently (dotted decimal vs. colon hex).   Separate IPv4-Net and IPv6-Net types are needed because the address portion is serialized differently and because the allowed values of the CIDR prefix are different (0-32 vs. 0-128).

**Comment by sparrell** [comment]
I think I agree with Dave K – at least with respect to what we agreed to. Personally I could go either way on the data type of a single ip address since it is not a use case I care about (I’d use a /32 in a net in a ipconn if I only needed a single addr). But what I remember from the meeting was Dave’s point on other potential serializations so I thought we agreed to a data type of an integer with a serialization of “1.2.3.4” per the appropriate two RFC’s (one for the ip, one for text representation). This is for the single addr case. For the ipnet case, I think the data type is a string of type “1.2.3.4/16”.

Duncan Sparrell
sFractal Consulting LLC
iPhone, iTypo, iApologize
I welcome VSRE emails. Learn more at http://vsre.info/

**Comment by davaya** [comment]
I also remember discussion at the F2F that OpenC2 should be standards-compliant.  (That was with regard to whether an IP address string with leading zeroes ("1.2.03.4") should or should not be considered valid.  The consensus was that OpenC2 should follow standards, not whatever common implementations do.)

An IPv4 Address is defined to be a 32 bit value in RFC 791 https://tools.ietf.org/html/rfc791#section-3.1.

Surprisingly, there is **no** standard for the text representation of an IPv4 address - the closest that exists is an Internet Draft that expired in 2005 - https://tools.ietf.org/html/rfc791#section-3.1.

There is an RFC for the text representation of IPv6 addresses (https://tools.ietf.org/html/rfc5952#section-4), but even that standard only makes recommendations, not requirements, for producers ("The recommendation in this section SHOULD be followed by systems when generating an address to be represented as text,"), and it explicitly requires consumers to accept a bunch of other text formats ("all implementations MUST accept and be able to handle any legitimate [RFC4291] format.").

Being standards-compliant means following the standards for what an IP address is (a 32 or 128 bit value), and how it is represented in text format (such as when using JSON serialization).

**Comment by davaya** [comment]
Based on the 1/29 LSC meeting, there are several separate questions to be decided.  The first is:

* Should there be separate v4 and v6 IP-related target types?

On the poll scale of 1: strongly prefer, 2: slightly prefer, 3: don't care, 4: slightly oppose, 5: strongly oppose, I'm a 2.   I believe it is cleaner for a receiver to know what type to expect before parsing it rather than having to parse the value to determine what type you received.

**Comment by davaya** [comment]
Third question:

* Should the type of IP-related targets be Binary or String?

This specification describes ONLY the JSON serialization, so IP-related targets are ALWAYS transmitted as strings (e.g., "192.168.130.27/24") regardless of the decision on this question.

But in the future OpenC2 will be used in constrained environments where sending IP addresses as text strings is inefficient in both bandwidth and processing.  Imagine if routers had to parse IP packets with addresses in string format instead of fixed 32 bit or 128 bit values.  The Type of the IP-related targets must be Binary in order for OpenC2 to be usable in both rich (JSON messages) and constrained (CBOR messages) environments.

**Comment by romanojd** [comment]
I'm going to bypass your questions, @davaya, and propose a pull request so we can discuss this with something tangible. I'll have it ready this afternoon.

**Comment by romanojd** [comment]
Pull request is ready. Have at it.

**Comment by davaya** [comment]
Comments left on #237 and summarized here.   The PR partially addresses the issue (creates separate IPv4 and IPv6) but creates other problems.

- The IETF defines an IPv4 Address in the IPv4 RFC: https://tools.ietf.org/html/rfc791#section-3.1.
- The IETF defines the "Text Representation of an IPv4 Address" in https://tools.ietf.org/html/draft-main-ipaddr-text-rep-00#section-3.1.
- The IETF defines CIDR and the rationale for netmasks in RFC 4632: ""Classless Inter-domain Routing (CIDR): The Internet Address Assignment and Aggregation Plan" https://tools.ietf.org/html/rfc4632#section-5

OpenC2 needs to be standards-compliant - that means a CIDR Address Range is not the same thing as an IP Address, and the Text Representation of an IP Address is not the same thing as an IP Address. The Pull Request needs to:
- Leave the definition of IPv4 and IPv6 Address as Binary, not change it to String
- Create new IPv4 and IPv6 Net targets that are an IP Address plus an optional CIDR prefix length
- Define JSON serializations for Text Representation of IP Address and IP Net
- Define IPv4 and IPv6 Connection targets that use IPv4 and IPv6 Net (CIDR) address ranges for src_addr and dst_addr.

**Comment by dlemire60** [comment]
I'm amazed that the best reference for textual representation of IP addresses is a 14 y.o. Internet Draft.

**Comment by davaya** [comment]
Per the 5 Feb LSC meeting, I do not have a strong opinion that both IP-Addr and IP-Net targets are needed for the one use case of the language we have defined today (the SLPF).   So if there is consensus for getting rid of the "single IP Address" target, I don't object to postponing it for lack of a defined use case.   I do want to reserve the term IP-Addr for when a "single IP Address" target will be needed in the future.

So if we delete IPv4-Addr and IPv6-Addr, then we need a name for the "one or more IP Addresses" target.   Those names are proposed in #253 to be IPv4-Net and IPv6-Net.  But the names could be anything *except* the unused, reserved "single IP address" name IPv4-Addr.   The names could be IPv4-Addrs (plural), or IPv4-Net, or IPv4-Addr-Range, whatever the LSC prefers.

**Comment by jmbrule** [comment]
Agree with postpone for a use case.  I do not have a strong opinion on what to name it.  If nobody else cares, then may I suggest IP-Net?  Again, select the name you like

**Comment by romanojd** [comment]
Dave, I see that you removed the "name" column from the address range datatypes, but I don't see why you need the second property that was named `prefix-len`. Let me try to connect the dots. To understand the target `ipv4_net`.

1. `ipv4_net` -> type: "IPv4-Net"
2. type: "IPv4-Net" ("IPv4 Address Range") -> type: "Array.ipv4-net", which contains two elements, type: "IPv4-Addr" and type: "Integer" for CIDR prefix-length
3. type: "IPv4-Addr" ("IPv4 Address") -> type: "Binary.ipv4-addr"
4. type: "Binary.ipv4-addr" -> serialized as JSON string

I understand that to represent a range you need an address and a prefix-length, but I don't think the Language Spec needs to define this. Why can't we simplify this further and go from 1 to 4 and be done?

**Comment by davaya** [comment]
At 2/12 LSC I took the action to 1) add explanatory text to the IPv*_Net target for how it is represented as a string, and 2) submit a Use Case for using serializations other than JSON for post-1.0.

**Comment by romanojd** [comment]
LSC Meeting, 3/12/2019: @sparrell and @davaya agreed to update pull request #253 with less explanatory information and more examples. We will plan to review this at the next LSC Meeting.

**Comment by romanojd** [comment]
LSC Meeting, 3/20/2019: There were no objections to apply pull request #253 and closing this issue.

**Comment by romanojd** [comment]
This pull request was inadvertently merged with the "master" branch instead of "working". I reset the "master" branch back to commit 00a0485. I attempted to re-apply this pull request to the correct "working" branch, but apparently, once the pull request is merged, it's also closed and cannot be re-opened.

**Comment by romanojd** [comment]
Pull request #289 was created to update the "working" branch with the original changes from #253.
